### PR TITLE
New Goodput Flags for flexibility

### PR DESCRIPTION
**Pull request by Perseus14**
# Description

- New flags `enable_gcp_goodput_metrics` and `enable_gcp_step_deviation_metrics` that publishes goodput and step time deviation data to GCM for flexibility
- Workaround for b/409244592 
- Updated relevant docs

# Tests

Logs: https://cloudlogging.app.goo.gl/uE6aEzjmPTkGxqJw6
E2E on TPU: https://screenshot.googleplex.com/BZrMNqyKApboqLM
# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/sft_trainer.py [hunk 1]
```diff
@@ -346,12 +357,13 @@ def main(argv: Sequence[str]) -> None:
...
         step_deviation_interval_seconds=config.step_deviation_interval_seconds,
+        gcp_options=gcp_options,
     )
     goodput_monitor.start_goodput_uploader()
-    max_logging.log("Started Goodput upload to Tensorboard in the background!")
+    max_logging.log("Started Goodput upload to Tensorboard & GCM in the background!")
     if config.monitor_step_time_deviation:
       goodput_monitor.start_step_deviation_uploader()
-      max_logging.log("Started step time deviation upload to Tensorboard in the background!")
+      max_logging.log("Started step time deviation upload to Tensorboard & GCM in the background!")
```
Comment (SurbhiJainUSC) [inline MaxText/sft_trainer.py:366]: Isn't this log misleading? We are turning off step time deviation metrics upload to GCM, right?

Comment (Perseus14) [reply MaxText/sft_trainer.py:366]: By default, we are enabling goodput and step deviation flags. If `report_performance_metric_for_gcp_monitoring` flag is set to `True` as per [Workload Observability](https://github.com/AI-Hypercomputer/maxtext/blob/main/getting_started/GCP_Workload_Observability.md), we disable sending step-deviation to GCM in order not to cause double publishing. As default value of `report_performance_metric_for_gcp_monitoring` is `False`, in most cases step deviation data is sent to GCM.

File: MaxText/train.py [hunk 2]
```diff
@@ -964,6 +964,17 @@ def main(argv: Sequence[str]) -> None:
 
   if config.monitor_goodput and jax.process_index() == 0:
     logger_name = f"goodput_{config.run_name}"
+    # Workload monitoring and Goodput monitoring both uses /workload/performance
+    # GCM metric to publish step_time and step_deviation metrics. For now, we
+    # will disable publishing step deviation metrics to GCM if workload
+    # monitoring is enabled. Will reconcile this in the future.
```
[the remark this answers is not in the document]
Comment (Perseus14) [reply MaxText/train.py:970]: [Workload Observability](https://github.com/AI-Hypercomputer/maxtext/blob/main/getting_started/GCP_Workload_Observability.md)

**Review by SurbhiJainUSC — APPROVED** [review]

File: MaxText/sft_trainer.py [hunk 3]
```diff
@@ -336,6 +336,17 @@ def main(argv: Sequence[str]) -> None:
 
   if config.monitor_goodput and jax.process_index() == 0:
     logger_name = f"goodput_{config.run_name}"
+    # Workload monitoring and Goodput monitoring both uses /workload/performance
+    # GCM metric to publish step_time and step_deviation metrics. For now, we
+    # will disable publishing step deviation metrics to GCM if workload
+    # monitoring is enabled. Will reconcile this in the future.
+    if config.report_performance_metric_for_gcp_monitoring:
```
Comment (dipannita08) [inline MaxText/sft_trainer.py:343]: I'm confused, there should be only one source (and exporter) of step time deviation and it should be the Goodput library, could you share more details on why this is required?

**Review by dipannita08 — CHANGES_REQUESTED** [review]

File: MaxText/sft_trainer.py (hunk 3 above)
Comment (Perseus14) [reply MaxText/sft_trainer.py:343]: Maxtext has a [feature](https://github.com/AI-Hypercomputer/maxtext/blob/main/getting_started/GCP_Workload_Observability.md) that publishes `step_time` to `/workload/performance` system metric. This is disabled by default. The new Goodput changes publishes `step_time_deviation` to the same system metric by default.

To ensure that there are no conflicts, when a flag is enabled to publish `step_time` we will disable publishing `step_time_deviation`.

File: getting_started/Monitor_Goodput.md [hunk 4]
```diff
@@ -151,52 +151,30 @@ goodput_monitor = monitoring.GoodputMonitor(
 If you do not wish to send metrics to Google Cloud Monitoring then please set
 the flag `enable_gcp_goodput_metrics` to `False` for disabling goodput metrics
 and `enable_gcp_step_deviation_metrics` to `False` for disabling step deviation
-metrics while creating the GCPOptions object.
+metrics.
 
-Setting `monitoring_enabled` to `False` will disable both tensorboard and GCM
```
Comment (dipannita08) [inline getting_started/Monitor_Goodput.md:156]: Why are we removing these instructions from the doc?

**Review by dipannita08 — APPROVED** [review]

**Review by bvandermoon — APPROVED** [review]

**Review by bvandermoon — APPROVED** [review]

**Review by bvandermoon — APPROVED** [review]
FYI - you will need to squash commits before it will let you merge
